File: 2020/2/2/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func readInput(filename string) ([]string, []string) {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	policies := make([]string, 0)
	passwords := make([]string, 0)
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := strings.Split(scanner.Text(), ":")
		policy, password := line[0], line[1]
		policies = append(policies, policy)
		passwords = append(passwords, password)
	}
	return policies, passwords
}

func isValid(policy string, password string) bool {
	p := strings.Split(policy, " ")
	count, character := p[0], p[1]
	c := strings.Split(count, "-")

	pos1, err := strconv.ParseInt(c[0],10 , 64)
	if err != nil {
		log.Fatal(err)
	}
	pos2, err := strconv.ParseInt(c[1], 10, 64)
	if err != nil {
		log.Fatal(err)
	}

	pass := strings.Split(password, "")
	if pass[pos1] == character && pass[pos2] != character {
		return true
	} else if pass[pos1] != character && pass[pos2] == character {
		return true
	} else {
		return false
	}
}

func calculate(policies []string, passwords []string) int {
	valid_passwords := 0
	for i, _ := range policies {
		if isValid(policies[i], passwords[i]) {
			valid_passwords += 1
		}
	}
	return valid_passwords
}

func main() {
	policies, passwords := readInput("input.txt")
	result:= calculate(policies, passwords)
	fmt.Println(result)
}

File: 2020/2/2/main.py
#!/usr/bin/env python3

import sys

def read_input(filename):
    policies = []
    passwords = []
    with open(filename) as fd:
        lines = fd.readlines()
        for line in lines:
            line = line.split(':')
            policies.append(line[0].strip())
            passwords.append(line[1].strip())
    return policies, passwords

def is_valid(policy, password):
    policy = policy.split()
    count, character = policy[0].split('-'), policy[1]
    pos1, pos2 = int(count[0]), int(count[1])
    if password[pos1-1] == character and password[pos2-1] != character:
        return True
    elif password[pos1-1] != character and password[pos2-1] == character:
        return True
    else:
        return False

def calculate(policies, passwords):
    valid_passwords = 0
    for i in range(0, len(policies)):
        if is_valid(policies[i], passwords[i]):
            valid_passwords += 1
    return valid_passwords

def main():
    policies, passwords = read_input("input.txt")
    result = calculate(policies, passwords)
    print(result)

if __name__ == '__main__':
    sys.exit(main())